module/apmprometheus/gatherer.go

```diff
@@ -80,8 +81,60 @@ func (g gatherer) GatherMetrics(ctx context.Context, out *apm.Metrics) error {
 					out.Add(name+".percentile."+strconv.Itoa(p), labels, q.GetValue())
 				}
 			}
+		case dto.MetricType_HISTOGRAM:
+			// For the bucket values, we follow the approach described by Prometheus's
+			// histogram_quantile function (https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile)
+			// to achieve consistent percentile aggregation results:
+			//
+			// "The histogram_quantile() function interpolates quantile values by assuming a linear
+			// distribution within a bucket. (...) If a quantile is located in the highest bucket,
+			// the upper bound of the second highest bucket is returned. A lower limit of the lowest
+			// bucket is assumed to be 0 if the upper bound of that bucket is greater than 0. In that
+			// case, the usual linear interpolation is applied within that bucket. Otherwise, the upper
+			// bound of the lowest bucket is returned for quantiles located in the lowest bucket."
+			for _, m := range mf.GetMetric() {
+				h := m.GetHistogram()
+				// Total count for all buckets in this
+				// histogram. We want the per bucket count.
+				if h.GetSampleCount() == 0 {
+					continue
+				}
+				labels := makeLabels(m.GetLabel())
+				buckets := h.GetBucket()
+				midpoints := make([]float64, len(buckets))
+				counts := make([]uint64, len(buckets))
+				for i, b := range buckets {
+					le := b.GetUpperBound()
+					if i == 0 {
+						if le > 0 {
+							le /= 2
+						}
+						// apm-server expects non-cumulative
+						// counts. prometheus counts each
+						// bucket cumulatively, ie. bucketN
+						// contains all counts for bucketN and
+						// all counts in preceding buckets. To
+						// get the current bucket's count we
+						// subtract bucketN-1 from bucketN,
+						// when N>0.
+						counts[i] = b.GetCumulativeCount()
+					} else if i == (len(buckets) - 1) {
+						le = buckets[i-1].GetUpperBound()
+						counts[i] = b.GetCumulativeCount() - buckets[i-1].GetCumulativeCount()
+					} else {
+						le = buckets[i-1].GetUpperBound() + (le-buckets[i-1].GetUpperBound())/2.0
+						counts[i] = b.GetCumulativeCount() - buckets[i-1].GetCumulativeCount()
+					}
+					// TODO: This rounds to 10,000ths, ie.
+					// 0.00001, precision, to handle
+					// floating point math trailing values,
+					// ie. 0.001500000003.
+					// How do we want to handle this?
+					midpoints[i] = math.Round(le*10000) / 10000
```

Comment:
I'm wondering if it's worth rounding/truncating at all. There might be legitimate use cases for indexing histograms with lower than 10000ths, and we'd be artificially excluding them. What problem are we solving with this? Just keeping the payload size down?

Reply:
When performing the midpoint math on buckets, I was getting bucket values like what's in the comment, like `0.001500000003`, which is probably not desired, and also might cause for flaky tests on different hardware (I think?).

Reply:
OK, I see. I think the fact that there's floating point error is fine and expected, but we should certainly avoid flaky tests. Could we find some bucket values which won't have this issue in tests?

Reply:
Updated, hopefully CI is also happy